pkg/cloud/externalconn/connectionpb/connection.proto

```diff
@@ -18,23 +18,35 @@ import "cloud/cloudpb/external_storage.proto";
 enum ConnectionProvider {
   Unknown = 0;
   nodelocal = 1;
+  gs_kms = 2;
 }
 
 enum ConnectionType {
   option (gogoproto.goproto_enum_prefix) = false;
 
   UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "TypeUnspecified"];
   STORAGE = 1 [(gogoproto.enumvalue_customname) = "TypeStorage"];
+  KMS = 2 [(gogoproto.enumvalue_customname) = "TypeKMS"];
 }
 
 message ConnectionDetails {
```

Comment:
Having a Details be completely unaware of its `ConnectionType` is starting to feel weird to me (mea culpa for not noticing this earlier) - it just seems kind of strange to have `nodelocal` and `gcs_kms` in the same list, when they're such radically different categories (as indicated by that `kms` suffix!). Think there's value in folding a Type into a ConnectionDetails, and then perhaps having a nested oneof's? I.e. something like
```
oneof detailsByType {
  StorageDetails storage = 1
  KMSDetails ks = 2
}
```

and then in `KMSDetails`: 
```
oneof details {
  GCSKMSConnectionDetails gcs_kms =1
  AWS... aws_kms=2
}
```
Or maybe that's too verbose for not enough gain? Or just wrong? Or difficult? We could also leave as-is, and just rename `nodelocal` to `storage_nodelocal` or some such (either by flat-out renaming, which is probably fine under the circumstances, or by deprecating field 2 and adding the new field as number 4)

Reply:
The reason for having a `ConnectionType` be an External Connection object level property (and a column in the system table) was to allow for tools like:

`SHOW BACKUP EXTERNAL CONNECTIONS` 

to be built in the future. These could be driven by delegate `SELECT` queries that filter out relevant rows without having to unmarshal each detail and look inside it. As you might've suggested before maybe we could add a method to `connection.go` that looks something like:

`func (d *ConnectionDetails) Type() {}` that returns the type based on the Details we give it. See `pkg/jobs/jobspb/wrap.go` I think jobs does a similar thing. That way we tie details to their types without actually storing them in the details proto. Wdyt?

Reply:
I am convinced that we want a database column with connection type in it!

I'm not convinced that we've nailed a strong polymorphic model here yet. Thinking about it, do we actually the different types of external connections to be the same object? They have a `Dial` API in common...only we don't want that `Dial` method to actually be cross-callable, which feels like the sort of thing a compiler can help with! But each external connection type is created differently, stored differently and used differently...do they have enough in common to justify a shared struct and proto? Very curious what @miretskiy thinks too!

Reply:
> But each external connection type is created differently, stored differently and used differently  

Isn't this what an interface is solving for us? We have different implementations of this interface that each know what its details are, how they are stored, what to do with them, what to redact and so on. As far as a shared proto goes, my understanding is that a `oneof` is appropriate for this use case where you have many independent protos, each with their own unique properties but belonging to the same umbrella of an `External Connection`. Jobs is similar in its layout where we have a top-level `Details` proto with several specialized `BACKUP, RESTORE,CDC` details nested in it. Each job decides how to implement resume, onfailorcancel etc.

Reply:
Not being able to call `Dial` on a KMS object when you actually want an ExternalStorage seems like a problem that can be solved by consolidating the codepath that reads these objects from the system table. We can come up with an API where you have to go to the method with a type you want, and the method will only return you the object if it is of that type.

Reply:
Discussed offline!
- We're pretty confident we won't need more than 2/3 connection types for a while, but not certain. That means we want to store all of these in the same database table - while it's scalable enough to add the occasional struct or interface, system tables are more heavyweight.
- Given that all of these external connection types are going in the same table, we need to use some kind of type-checking/casting to convert the row field into the appropriate type of object. This can and will happen exclusively in the external connection library.
- Given _that_, having multiple interfaces doesn't buy us anything. So we probably also want to leave a single top-level proto for consistency.